matrix_sdk_appservice/tests/tests.rs

```diff
@@ -28,33 +39,222 @@ async fn appservice(registration: Option<Registration>) -> Result<Appservice> {
     let homeserver_url = mockito::server_url();
     let server_name = "localhost";
 
-    Ok(Appservice::new(homeserver_url.as_ref(), server_name, registration).await?)
+    let client_config =
+        ClientConfig::default().request_config(RequestConfig::default().disable_retry());
+
+    Ok(Appservice::new_with_config(
+        homeserver_url.as_ref(),
+        server_name,
+        registration,
+        client_config,
+    )
+    .await?)
 }
 
-fn member_json() -> serde_json::Value {
-    json!({
-        "content": {
-            "avatar_url": null,
-            "displayname": "example",
-            "membership": "join"
-        },
-        "event_id": "$151800140517rfvjc:localhost",
-        "membership": "join",
-        "origin_server_ts": 151800140,
-        "room_id": "!ahpSDaDUPCCqktjUEF:localhost",
-        "sender": "@example:localhost",
-        "state_key": "@example:localhost",
-        "type": "m.room.member",
-        "prev_content": {
-            "avatar_url": null,
-            "displayname": "example",
-            "membership": "invite"
-        },
-        "unsigned": {
-            "age": 297036,
-            "replaces_state": "$151800111315tsynI:localhost"
-        }
-    })
+#[async_test]
+async fn test_register_virtual_user() -> Result<()> {
+    let appservice = appservice(None).await?;
+
+    let localpart = "someone";
+    let _mock = mockito::mock("POST", "/_matrix/client/r0/register")
+        .match_query(mockito::Matcher::Missing)
+        .match_header(
+            "authorization",
+            mockito::Matcher::Exact(format!("Bearer {}", appservice.registration().as_token)),
+        )
+        .match_body(mockito::Matcher::Json(json!({
+            "username": localpart.to_owned(),
+            "type": "m.login.application_service"
+        })))
```

Comment:
Try intermediate conversion from `http::Request` instead (see https://github.com/ruma/ruma/issues/610#issuecomment-854558942)

Reply:
Moved to tracking issue